proposals/0003-spec.md

```diff
@@ -0,0 +1,249 @@
+* MUST be appropriate for machine-to-machine communication
+* SHOULD give enough information for a human administrator
+
+
+### Terms used
+
+| Term       | Description     |
+| ------------- |-------------|
+| Producer | The service/application that is checked |
+| Consumer | The probing end, usually a machine, that needs to verify the liveness of a Producer |
+| Health Check Procedure | The code executed to determine the liveliness of a Producer |
+| Producer Outcome | The overall outcome, determined by considering all health check procedure results |
+| Health check procedure result | The result of single check |
+
+# Protocol Overview
+
+
+1. Consumer invokes the health check of a Producer through any of the supported protocols
+2. Producer enforces security constraints on the invocation (i.e authentication)
+3. Producer executes a set of Health check procedures (could be a set with one element)
+4. Producer determines the overall outcome (Producer outcome)
+5. The outcome is mapped to outermost protocol (i.e. HTTP status codes)
+6. The payload is written to the response stream
+7. The consumer reads the response
+8. The consumer determines the overall outcome
+Protocol Specifics
+Interacting with producers
+How are the health checks accessed and invoked ? We don’t make any assumptions about this, except for the wire format and protocol.
+
+## Protocol Mappings
+
+* Producers MAY support a variety of protocols but the information items in the response payload MUST remain the same.
+* Producers SHOULD define a well known default context to perform checks
+Protocol Mappings
+Health checks (innermost) can and should be mapped to the actual invocation protocol (outermost). This section described some of guidelines and rules for these mappings.
+* Each response SHOULD integrate with the outermost protocol whenever it makes sense (i.e. using HTTP status codes to signal the overall state)
+* Inner protocol information items MUST NOT be replaced by outer protocol information items, rather kept redundantly.
+* The inner protocol response MUST be self-contained, that is carrying all information needed to reason about the the producer outcome
+
+
+## Mandatory and optional protocol types
+REST/HTTP interaction
+* Producer MUST provide a HTTP endpoint that follow the REST interface specifications described in Appendix A
+
+
+## Protocol Adaptor
+
+
+Each provider MUST provide the REST/HTTP interaction, but MAY provide other protocols such as TCP or JMX. When possible, the output MUST be the JSON output returned by the equivalent HTTP calls (Appendix B). The request is protocol specific.
+Healthcheck Response information
+
+
+* The primary information MUST be boolean, it needs to be consumed by other machines. Anything between available/unavailable doesn’t make sense or would increase the complexity on the side of the consumer processing that information.
```

Comment:
I think we should definitely require a helthcheck to give a binary status UP/DOWN to make things easy (the `outcome` field). But we may specify optional outcomes to support shades of gray, e.g. a service would return `outcome=UP`, but can and additional (optional) fields like condition (percentual status between 0 (all bad) and 100 (all good), or status (GREEN, YELLOW, RED, WARNINGS, etc.) However, this can be added later if we don't find a consensus soon.